### PR TITLE
fix(forward-date): past times now increment day by one

### LINKED ISSUE #455: Bug: forwardDate fails if the specified time is passed the reference time
## Abstract

If the current time is `1:57 AM`, passing `"1am"` with `forwardDate: true` results in date referencing `1am` on the _current date._

Instead, since `1:00 AM` has already passed by 57 minutes, the result should be `1am` the following day.

Realistically, `forwardDate: true` should never result in a date that is older than the reference date.

## Example

**Reference Date:** May 26th at `01:57 AM`
**Input**: `"1am"`
**Timezone:** `CDT`
```javascript
const now = new Date(); // 2022-05-26T06:57:55.240Z

chrono.parseDate("1am", now, { forwardDate: true }); // 2022-05-26T06:00:00.000Z
```

**Expected:** `2022-05-27T06:00:00.000Z` (May 27 @ 1am)
**Actual:** `2022-05-26T06:00:00.000Z` (May 26 @ 1am)

## Addendum

Interestingly enough, now that it is `2:00AM`, Chrono still insists on using the reference day, instead of the following day.

```javascript
const now = new Date(); // 2022-05-26T07:14:28.103Z
chrono.parseDate('1am', now, { forwardDate: true }); // 2022-05-26T06:00:00.000Z
```

### PR DESCRIPTION
**Pull request by Swivelgames**
Fixes #455

**_This PR introduces and successfully passes its tests._**

| Note: Unrelated tests were failing locally at the time that this repo was cloned.<br><br>Isolating and removing the changes in this PR did not affect the outcome of the previously failing tests, suggesting that these changes do not affect them.<br><br>_This branch needs to be pulled down and tested locally by a maintainer before it is considered for merging._ |
| --- |

## Build
signal-2022-05-26-041047_001

## Tests
signal-2022-05-26-041047_002

### CODE REVIEW

File: src/common/refiners/ForwardDateRefiner.ts [hunk 1]
```diff
@@ -17,6 +17,27 @@ export default class ForwardDateRefiner implements Refiner {
...
+                    result.end.imply("day", refMoment.date());
+                    result.end.imply("month", refMoment.month() + 1);
+                    result.end.imply("year", refMoment.year());
+
+                    if (refMoment.isAfter(result.end.dayjs())) {
+                        refMoment = refMoment.add(1, "day");
+                        result.end.imply("day", refMoment.date());
+                        result.end.imply("month", refMoment.month() + 1);
+                        result.end.imply("year", refMoment.year());
+                    }
```
Comment (Swivelgames) [inline src/common/refiners/ForwardDateRefiner.ts:37]: This seems redundant, but it prevents a scenario where the end date lands on the same day that we've adjusted the start date to -- Specifically when the end _**time**_ is earlier in the day then the **_start time_**.

File: test/en/en_time_exp.test.ts [hunk 2]
```diff
@@ -287,3 +287,84 @@ test("Test - Parsing negative cases : 'at [some numbers] - [some numbers]' (Stri
...
+            expect(result.index).toBe(2);
+            expect(result.text).toBe("1am");
+
+            expect(result.start).not.toBeNull();
+            expect(result.start.get("year")).toBe(2022);
+            expect(result.start.get("month")).toBe(5);
+            expect(result.start.get("day")).toBe(27);
+            expect(result.start.get("hour")).toBe(1);
+
+            expect(result.start).toBeDate(new Date(2022, 5 - 1, 27, 1));
```
Comment (wanasit) [inline test/en/en_time_exp.test.ts:312]: Nit: Recently, I think testing the date is redundant to testing date and hour about.

File: src/common/refiners/ForwardDateRefiner.ts (hunk 1 above)
Comment (wanasit) [reply src/common/refiners/ForwardDateRefiner.ts:37]: Should we compare `refMoment.isAfter(result.end.dayjs())`? or `result.start.dayjs().isAfter(result.end.dayjs())`?

I think the latter is more robust??

File: src/common/refiners/ForwardDateRefiner.ts [hunk 3]
```diff
@@ -17,6 +17,27 @@ export default class ForwardDateRefiner implements Refiner {
         results.forEach(function (result) {
             let refMoment = dayjs(context.refDate);
 
+            if (result.start.isOnlyTime() && refMoment.isAfter(result.start.dayjs())) {
+                refMoment = refMoment.add(1, "day");
+
+                result.start.imply("day", refMoment.date());
```
Comment (wanasit) [inline src/common/refiners/ForwardDateRefiner.ts:23]: Nit: Could you add `implySimilarDate()` (see. `assignSimilarDate()`) in the `util/dayjs` and use the function here?

File: README.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # Chrono (v2)
 
-A natural language date parser in Javascript.
```
Comment (wanasit) [inline README.md:3]: What autoformat do you use? Could you revert the style change on this file? Or submit it separately with the suggested formatter?

**Review by wanasit — COMMENTED** [review]
Hello @Swivelgames. Thank you very much for the improvement!

Sorry for my slow reply. Could you please take a look at my comments?

File: README.md (hunk 4 above)
Comment (Swivelgames) [reply README.md:3]: Woops! My bad! :smile: -- I just format my files with `sed`, and I guess it picked up the `README.md` on accident. Reverted! I'll push it up shortly.

>```
>$ sed -i'' -E 's,[[:space:]],
>```

**Comment by Slapbox** [comment]
@wanasit @Swivelgames thanks both for your hard work!

Is there anything left to do to get this merged besides make those few requested changes?

image

**Comment by wanasit** [comment]
@Swivelgames I have incorporated your change into 996a99872eddc587fbc57e0476bc3012dfdc292d.
So closing this.

**Comment by Swivelgames** [comment]
@Slapbox @wanasit Apologies for the hiatus! I'm so glad you were able to incorporate these changes in. Sorry for disappearing! I was going through a pretty major life transition :slightly_smiling_face:

**Pull request closed without merging**